File: backend/src/memory/memory.service.ts
import { Injectable } from '@nestjs/common'
import { HumanMessage, AIMessage, BaseMessage } from '@langchain/core/messages'

@Injectable()
export class MemoryService {
  private sessions = new Map<string, BaseMessage[]>()
  private readonly MAX_HISTORY = 20

  getHistory(userId: string): BaseMessage[] {
    return this.sessions.get(userId) || []
  }

  addMessage(userId: string, message: BaseMessage) {
    const history = this.sessions.get(userId) || []
    history.push(message)
    if (history.length > this.MAX_HISTORY) history.splice(0, history.length - this.MAX_HISTORY)
    this.sessions.set(userId, history)
  }

  clearHistory(userId: string) { this.sessions.delete(userId) }

  listSessions() {
    return Array.from(this.sessions.entries()).map(([id, msgs]) => ({ userId: id, messageCount: msgs.length }))
  }
}

File: backend/app/memory/memory_service.py
from langchain_core.messages import BaseMessage

MAX_HISTORY = 20

_sessions: dict[str, list[BaseMessage]] = {}


def get_history(user_id: str) -> list[BaseMessage]:
    return _sessions.get(user_id, [])


def add_message(user_id: str, message: BaseMessage) -> None:
    history = _sessions.get(user_id, [])
    history.append(message)
    if len(history) > MAX_HISTORY:
        del history[: len(history) - MAX_HISTORY]
    _sessions[user_id] = history


def clear_history(user_id: str) -> None:
    _sessions.pop(user_id, None)


def list_sessions() -> list[dict]:
    return [{"userId": uid, "messageCount": len(msgs)} for uid, msgs in _sessions.items()]
